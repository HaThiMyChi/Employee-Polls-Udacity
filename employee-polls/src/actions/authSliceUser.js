import { SIGNOUT_AUTH_SLICE_USER, SET_AUTH_SLICE_USER } from "../constant/constant";

export function setAuthSliceUser(authedUser) {
    return {
        type: SET_AUTH_SLICE_USER,
        authedUser
    };
}

export function signoutAuthSliceUser() {
    return {
        type: SIGNOUT_AUTH_SLICE_USER
    };
}

export function handleLoginUser(username, password) {
    return (dispatch, getState) => {
        console.log('dispatch', dispatch)
        console.log('getState', getState)
        const { users } = getState();

        const user = Object.values(users).find((user) => user.id === username && user.password === password);

        if (!user) {
            return dispatch(setAuthSliceUser(user));
        }
    }
}
