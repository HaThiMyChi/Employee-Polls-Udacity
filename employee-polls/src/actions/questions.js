import { ADD_QUESTION, ADD_ANSWER_QUESTION, RECEIVE_QUESTIONS } from "../constant/constant";

export function receiveQuestions(questions) {
    return {
        type: RECEIVE_QUESTIONS,
        questions
    };
}

export function addQuestion(question) {
    return {
        type: ADD_QUESTION,
        question
    }
}

export function addAnswerQuestion(author, qid, answer) {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer
    }
}