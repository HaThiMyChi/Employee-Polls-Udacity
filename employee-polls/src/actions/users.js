import { RECEIVE_USERS } from "../constant/constant";

export function receiveUsers(users) {
    return {
        type: RECEIVE_USERS,
        users
    };
}