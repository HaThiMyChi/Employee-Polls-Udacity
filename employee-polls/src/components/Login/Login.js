import React from 'react';
import PropTypes from 'prop-types';

Login.propTypes = {

};

function Login(props) {
    return (
        <div>
            <h1 className='font-bold mt-9 text-3xl  text-center' data-testid="login-heading">
                Login
            </h1>
            <form>
                <div>
                    <label htmlFor='username' className='text-sm font-medium block text-slate-700'>Username</label>
                    <div className='mt-1'>
                        <input
                            type="text"
                            name="username"
                            id="username"
                            data-testid="username"
                            className='px-3 py-2 border bg-white shadow-sm border-slate-300 placeholder-slate-400 
                                disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
                                focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1
                                invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none'
                        />
                    </div>
                </div>

                <div className='mt-6'>
                    <label htmlFor='password' className='text-sm font-medium block text-slate-700'>Password</label>
                    <div className='mt-1'>
                        <input
                            type="password"
                            name="password"
                            id="password"
                            data-testid="password"
                            className='px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400
                            disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
                            focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1 
                            invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none'
                        />
                    </div>
                </div>
                <div className='mt-6 text-center'>
                    <button className='bg-sky-500 hover:bg-sky-700 px-5 py-2.5 text-sm leading-5 font-semibold rounded-md text-white' type="submit" data-testid="submit">Login</button>
                </div>
            </form>
        </div>

    );
}

export default Login;