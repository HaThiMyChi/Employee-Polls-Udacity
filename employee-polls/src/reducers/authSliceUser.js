import { SET_AUTH_SLICE_USER, SIGNOUT_AUTH_SLICE_USER } from "../constant/constant";

export default function authSliceUser(state = null, action) {
    switch (action.type) {
        case SET_AUTH_SLICE_USER:
            return action.authedUser;
        case SIGNOUT_AUTH_SLICE_USER:
            return null;
        default:
            return state;
    }
}