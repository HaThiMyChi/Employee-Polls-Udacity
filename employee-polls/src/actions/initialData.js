import { getUserQuestionData } from "../utils/api";
import { receiveQuestions } from "./questions";
import { receiveUsers } from "./users";

export function handleInitialData() {
    return (dispatch) => {
        return getUserQuestionData().then(({ users, questions }) => {
            dispatch(receiveUsers(users));
            dispatch(receiveQuestions(questions));
        });
    };
}