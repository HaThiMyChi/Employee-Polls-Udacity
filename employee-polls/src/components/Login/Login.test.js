import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { act } from 'react';
import Login from './Login';
import { handleLoginUser } from '../../actions/authSliceUser';

jest.mock('../../actions/authSliceUser', () => ({
    handleLoginUser: jest.fn(() => ({ type: 'LOGIN_USER' })),
}));

const mockStore = configureStore([]);

describe('Login', () => {
    let store;

    beforeEach(() => {
        store = mockStore({
            authSliceUser: null,
        });

        store.dispatch = jest.fn();
    });

    it('should render Login component with default values', () => {
        let component;
        act(() => {
            component = render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Login />
                    </MemoryRouter>
                </Provider>
            );
        });

        expect(screen.getByTestId('login-heading')).toBeInTheDocument();
        expect(screen.getByTestId('username')).toHaveValue('sarahedo');
        expect(screen.getByTestId('password')).toHaveValue('password123');

        expect(component.asFragment()).toMatchSnapshot();
    });

    it('should handle username and password change', () => {
        let component;
        act(() => {
            component = render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Login />
                    </MemoryRouter>
                </Provider>
            );
        });

        const usernameInput = screen.getByTestId('username');
        const passwordInput = screen.getByTestId('password');

        fireEvent.change(usernameInput, { target: { value: 'newUser' } });
        fireEvent.change(passwordInput, { target: { value: 'newPassword' } });

        expect(usernameInput).toHaveValue('newUser');
        expect(passwordInput).toHaveValue('newPassword');

        expect(component.asFragment()).toMatchSnapshot();
    });

    it('should dispatch login action on form submit', () => {
        let component;
        act(() => {
            component = render(
                <Provider store={store}>
                    <MemoryRouter>
                        <Login />
                    </MemoryRouter>
                </Provider>
            );
        });

        const submitButton = screen.getByTestId('submit');

        fireEvent.click(submitButton);

        expect(store.dispatch).toHaveBeenCalledWith(handleLoginUser('sarahedo', 'password123'));
        expect(screen.getByTestId('username')).toHaveValue('');
        expect(screen.getByTestId('password')).toHaveValue('');

        expect(component.asFragment()).toMatchSnapshot();
    });

    it('should redirect if already logged in', () => {
        store = mockStore({
            authSliceUser: { id: 'user1', name: 'John Doe' },
        });

        let component;
        act(() => {
            component = render(
                <Provider store={store}>
                    <MemoryRouter initialEntries={['/login?redirectTo=/dashboard']}>
                        <Routes>
                            <Route path="/login" element={<Login />} />
                            <Route path="/dashboard" element={<div>Dashboard</div>} />
                        </Routes>
                    </MemoryRouter>
                </Provider>
            );
        });

        expect(screen.queryByTestId('login-heading')).not.toBeInTheDocument();
        expect(screen.getByText('Dashboard')).toBeInTheDocument();

        expect(component.asFragment()).toMatchSnapshot();
    });
});
