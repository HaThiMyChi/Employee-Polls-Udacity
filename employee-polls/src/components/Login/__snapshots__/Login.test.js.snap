// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should dispatch login action on form submit 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="font-bold mt-9 text-3xl  text-center"
      data-testid="login-heading"
    >
      Login
    </h1>
    <form>
      <div>
        <label
          class="text-sm font-medium block text-slate-700"
          for="username"
        >
          Username
        </label>
        <div
          class="mt-1"
        >
          <input
            class="px-3 py-2 border bg-white shadow-sm border-slate-300 placeholder-slate-400 
 disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1
 invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none"
            data-testid="username"
            id="username"
            name="username"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="mt-6"
      >
        <label
          class="text-sm font-medium block text-slate-700"
          for="password"
        >
          Password
        </label>
        <div
          class="mt-1"
        >
          <input
            class="px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400
 disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1 
 invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none"
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="mt-6 text-center"
      >
        <button
          class="bg-sky-500 hover:bg-sky-700 px-5 py-2.5 text-sm leading-5 font-semibold rounded-md text-white"
          data-testid="submit"
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Login should handle username and password change 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="font-bold mt-9 text-3xl  text-center"
      data-testid="login-heading"
    >
      Login
    </h1>
    <form>
      <div>
        <label
          class="text-sm font-medium block text-slate-700"
          for="username"
        >
          Username
        </label>
        <div
          class="mt-1"
        >
          <input
            class="px-3 py-2 border bg-white shadow-sm border-slate-300 placeholder-slate-400 
 disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1
 invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none"
            data-testid="username"
            id="username"
            name="username"
            type="text"
            value="newUser"
          />
        </div>
      </div>
      <div
        class="mt-6"
      >
        <label
          class="text-sm font-medium block text-slate-700"
          for="password"
        >
          Password
        </label>
        <div
          class="mt-1"
        >
          <input
            class="px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400
 disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1 
 invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none"
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value="newPassword"
          />
        </div>
      </div>
      <div
        class="mt-6 text-center"
      >
        <button
          class="bg-sky-500 hover:bg-sky-700 px-5 py-2.5 text-sm leading-5 font-semibold rounded-md text-white"
          data-testid="submit"
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Login should redirect if already logged in 1`] = `
<DocumentFragment>
  <div>
    Dashboard
  </div>
</DocumentFragment>
`;

exports[`Login should render Login component with default values 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="font-bold mt-9 text-3xl  text-center"
      data-testid="login-heading"
    >
      Login
    </h1>
    <form>
      <div>
        <label
          class="text-sm font-medium block text-slate-700"
          for="username"
        >
          Username
        </label>
        <div
          class="mt-1"
        >
          <input
            class="px-3 py-2 border bg-white shadow-sm border-slate-300 placeholder-slate-400 
 disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1
 invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none"
            data-testid="username"
            id="username"
            name="username"
            type="text"
            value="sarahedo"
          />
        </div>
      </div>
      <div
        class="mt-6"
      >
        <label
          class="text-sm font-medium block text-slate-700"
          for="password"
        >
          Password
        </label>
        <div
          class="mt-1"
        >
          <input
            class="px-3 py-2 bg-white border shadow-sm border-slate-300 placeholder-slate-400
 disabled:bg-slate-50 disabled:text-slate-500 disabled:border-slate-200 
 focus:outline-none focus:border-sky-500 focus:ring-sky-500 block w-full rounded-md sm:text-sm focus:ring-1 
 invalid:border-pink-500 invalid:text-pink-600 focus:invalid:border-pink-500 focus:invalid:ring-pink-500 disabled:shadow-none"
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value="password123"
          />
        </div>
      </div>
      <div
        class="mt-6 text-center"
      >
        <button
          class="bg-sky-500 hover:bg-sky-700 px-5 py-2.5 text-sm leading-5 font-semibold rounded-md text-white"
          data-testid="submit"
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
